File: 5-longest-palindromic-substring/longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s.length() <= 1) {
            return s;
        }

        StringBuilder sb = new StringBuilder("#");
        for (int i =0;i<s.length();i++) {
            sb.append(s.charAt(i)).append("#");
        }
        String temp = sb.toString();
        int n = temp.length();
        int[] z = new int[n];

        int left = 0;
        int right = 0;

        for (int i = 1; i<n;i++) {
            z[i] =  i < right ? Math.min(right-i, z[left + (right-i)]) : 0;
            while (i + z[i] + 1 < n && i - z[i] - 1 >=0 && 
                    temp.charAt(i + z[i] + 1) == temp.charAt(i - z[i] - 1)) {
                z[i]++;
            }
            if ( i + z[i] > right ) {
                left =  i - z[i];
                right = i + z[i];
            }
        }

        int res = 1;
        int index = 0;
        for (int i = 0;i<n;i++) {
            if (res < z[i]) {
                res = z[i];
                index = (i - z[i])/2;
            }
        }

        return s.substring(index, index + res);
        
    }

    // public String longestPalindrome(String s) {
    //     if (s.length() <= 1) {
    //         return s;
    //     }

    //     int maxLen = 1;
    //     String maxStr = s.substring(0, 1);
    //     s = "#" + s.replaceAll("", "#") + "#";
    //     int[] dp = new int[s.length()];
    //     int center = 0;
    //     int right = 0;

    //     for (int i = 0; i < s.length(); i++) {
    //         if (i < right) {
    //             dp[i] = Math.min(right - i, dp[2 * center - i]);
    //         }

    //         while (i - dp[i] - 1 >= 0 && i + dp[i] + 1 < s.length() && s.charAt(i - dp[i] - 1) == s.charAt(i + dp[i] + 1)) {
    //             dp[i]++;
    //         }

    //         if (i + dp[i] > right) {
    //             center = i;
    //             right = i + dp[i];
    //         }

    //         if (dp[i] > maxLen) {
    //             maxLen = dp[i];
    //             maxStr = s.substring(i - dp[i], i + dp[i] + 1).replaceAll("#", "");
    //         }
    //     }

    //     return maxStr;
    // }
}

File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1:
            return s
        
        temp = "#" + "#".join(s) + "#"
        n = len(temp)
        z = [0 for _ in range(n)]
        left = 0
        right = 0

        for index in range(1, n):
            if index < right:
                z[index] = min(right-index, z[left + (right - index)]) # Left Mirror or distance from right bound
            else:
                z[index] = 0
            
            # Expand current window
            while (index - z[index] - 1) >=0  and (index + z[index] + 1) < n and temp[index - z[index] - 1] == temp[index + z[index] + 1]:
                z[index] += 1

            # Update Bounds
            if index + z[index] > right:
                right = index + z[index]
                left = index - z[index]
        
        res = 1
        start = 0
        for index in range(n):
            l = z[index]
            if l > res:
                res = l
                start = (index - l)//2
        
        return s[start: start+res]





        

        